scalafmt-sbt/src/main/scala/org/scalafmt/sbt/ScalafmtPlugin.scala

```diff
@@ -1,25 +1,169 @@
+      InputKey[Unit]("scalafmt-only", "Format only given file")
+  }
+  import autoImport._
+
+  private lazy val scalaConfig =
+    scalafmtConfig.map(Config.fromHoconFile(_).get)
+  private lazy val sbtConfig =
+    scalaConfig.map(conf => conf.copy(runner = conf.runner.forSbt))
+
+  private type Input = String
+  private type Output = String
+
+  private def withFormattedSources[T](
+      sources: Seq[File],
+      config: ScalafmtConfig
+  )(
+      onError: (File, Throwable) => T,
+      onFormat: (File, Input, Output) => T
+  ): Seq[Option[T]] = {
+    sources.map(
+      file => {
+        val input = IO.read(file)
+        val output = Scalafmt.format(input, config)
+
+        output match {
+          case Formatted.Failure(e) =>
+            if (config.runner.fatalWarnings) {
+              throw e
+            } else if (config.runner.ignoreWarnings) {
+              // do nothing
+              None
+            } else {
+              Some(onError(file, e))
+            }
+          case Formatted.Success(code) =>
+            Some(onFormat(file, input, code))
+        }
       }
+    )
   }
-  override def globalSettings: Seq[Def.Setting[_]] =
-    Seq(
-      commands += autoImport.scalafmt
-    ) ++
-      addCommandAlias("scalafmtTest", "scalafmt --test") ++
-      addCommandAlias("scalafmtDiffTest", "scalafmt --diff --test") ++
-      addCommandAlias("scalafmtDiff", "scalafmt --diff")
 
+  private def formatSources(
+      sources: Seq[File],
+      config: ScalafmtConfig,
+      log: Logger
+  ): Unit = {
+    withFormattedSources(sources, config)(
+      (file, e) => log.error(s"Error in ${file.toString}: $e"),
+      (file, input, output) => {
+        if (input != output) {
+          log.info(s"Successfully formatted ${file.toString}.")
+          IO.write(file, output)
+        }
+      }
+    )
+  }
+
+  private def checkSources(
+      sources: Seq[File],
+      config: ScalafmtConfig,
+      log: Logger
+  ): Boolean = {
+    withFormattedSources(sources, config)(
+      (file, e) => {
+        log.error(s"Error in ${file.toString}: $e")
+        false
+      },
+      (file, input, output) => {
+        val diff = input != output
+        if (diff) {
+          log.error(s"${file.toString} isn't formatted properly!")
+        }
+        diff
+      }
+    ).flatten.forall(x => x)
+  }
+
+  private lazy val sbtSources = thisProject.map(
+    proj => {
+      val rootSbt =
+        BuildPaths.configurationSources(proj.base).filterNot(_.isHidden)
+      val projectSbt =
+        (BuildPaths.projectStandard(proj.base) * GlobFilter("*.sbt")).get
+          .filterNot(_.isHidden)
+      rootSbt ++ projectSbt
+    }
+  )
+  private lazy val projectSources = thisProject.map(proj =>
+    (BuildPaths.projectStandard(proj.base) * GlobFilter("*.scala")).get)
+
+  lazy val scalafmtSettings: Seq[Def.Setting[_]] = Seq(
+    scalafmtOnCompile := false,
+    scalafmt := formatSources(
+      unmanagedSources.value,
```

Comment:
You should read `(unmanagedSources in scalafmt).value` instead, so that it is possible to override the set of files processed by scalafmt independently of the sources given to the compiler. In complex builds, this can be necessary.

Reply:
I think it's possible to do something like `unmanagedSources.in(scalafmt).or(unmanagedSources).value` so that it defaults to `unmanagedSources` if it's not defined when scoped under the `scalafmt` task.

Reply:
`(unmanagedSources in scalafmt).value` automatically defaults to `unmanagedSources.value` if the former is not defined. You don't need to do anything for that, it's built in the scopes mechanism.